### PR TITLE
Robust js interop

### PR DESCRIPTION
**Pull request by Joelius300**
This PR improves the interop-layer to work with delegates instead of bare static- and instance-methods which allows for much more flexibility (but maybe worse performance. I don't expect it but tests may be required).

On the javascript-side, the interop is much more flexible and reusable because it uses `...args` which allows for any number of arguments. If we didn't have this approach we'd need tests and special cases for all different callback-types which would be horrible.  
On the c#-side, it's gotten more restrictive in a good way (following the type-safety of c#). You now have a fixed c# delegate which declares the signature of the callback. This allows you to only pass in correct delegates and if you use a javascript-handler you can immediately see what arguments you can expect. In many cases, we just use `JsonElement` since it's hard to model the actual type that's returned or it just hasn't been done yet. But wherever possible we use ~`System.Text.Json`~ *nope, we should use the same thing for both serializing and deserializing which currently is `Json.Net`* to deserialize to the actual type, meaning that we check the delegate parameters and if it's not `object` or `JsonElement` we deserialize it in order to invoke the delegate with actual types.

This PR is a draft because we could actually use `Json.Net` for the deserialization part which would make the converters much easier probably and I did not realize that before (see b4883b7 and it's comment). Therefore I'll probably play around with that some more. Also `ObjectEnum` can't be read currently, so that's also something that needs to be done before merging.

Another point which I don't like about the PR currently is the sample for the legend callbacks. While it demostrates how to use it, you can see that it has really bad performance. Unlike with other c# callbacks (see the other samples), this one somehow has big performance issues (maybe because of the many invokes but that seems unlikely given how many times the `OnHover` callback is executed without lag).

One more issue is support for value callbacks on server-side. The dotnet dispatcher from javascript only allows async calls to c# on the server-side variant which means that it will return a promise and that is of course not the value we want to give to chart.js. The client-side dispatcher allows sync calls so we can pass the actual value to chart.js and it works. One thing we could do for this is ask the asp .net team if they will ever allow sync calls to c# in server-side blazor. A [SO question](https://stackoverflow.com/questions/59543973/use-async-function-when-consumer-doesnt-expect-a-promise) has already been asked but more or less only confirmed my worries.

When we split the projects (#78), we will probably keep `IMethodHandler` in ChartJs.CSharp but move the `DelegateHandler` and the `JavaScriptHandler` to ChartJs.Blazor in order to minimize the required dependencies.

This PR has a few side-effects like renaming `xAxes` to `XAxes` in c# and adding/removing a few properties which were not yet present or wrong.

I don't know if adding more callbacks should be part of this PR, maybe someone has an opinion on this. Adding them in this PR would give more room for tests but I think it might be better to keep this PR at a minimum and make it extensible(!!!).

### CODE REVIEW

File: src/ChartJs.Blazor/ChartJS/Common/Handlers/DelegateHandler.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    public class DelegateHandler<T> : IMethodHandler<T>, IDisposable
+        where T : Delegate
+    {
+        private const string DelegateMethodName = "Invoke";
+        private static readonly JsonSerializerOptions s_deserializeOptions = new JsonSerializerOptions
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+        };
+
+        private readonly ParameterInfo[] _delegateParameters;
```
Comment (Joelius300) [inline src/ChartJs.Blazor/ChartJS/Common/Handlers/DelegateHandler.cs]: TODO Make static (since it isn't shared amongst different generic "version")

File: src/ChartJs.Blazor/Util/ExpandoObjectExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+            string[] segments = path.Split('.');
+            IDictionary<string, object> source = value;
+            for (int i = 0; i < segments.Length; i++)
+            {
+                string segment = segments[i];
+                IDictionary<string, object> newSource;
+                try
+                {
+                    newSource = source[segment] as IDictionary<string, object>;
```
Comment (Joelius300) [inline src/ChartJs.Blazor/Util/ExpandoObjectExtensions.cs]: Use `ContainsKey`

File: src/ChartJs.Blazor/ChartJS/ChartJsInterop.cs [hunk 3]
```diff
@@ -107,46 +107,107 @@ public static ValueTask<bool> UpdateChart(this IJSRuntime jsRuntime, ConfigBase
...
-                clearConfigExpando.options = clearConfigExpando.options ?? new { };
-                clearConfigExpando.options.legend = clearConfigExpando.options.legend ?? new { };
-                clearConfigExpando.options.legend.onClick = dynamicChartConfig.Options.Legend.OnClick;
+                const string path = "options.onClick";
+                cleanChartConfig.SetValue(path, chartOnClick);
             }
 
-            if (dynamicChartConfig?.Options?.Legend?.OnHover != null
-                && dynamicChartConfig?.Options?.Legend?.OnHover is IHoverHandler)
+            if (dynamicChartConfig?.Options?.OnHover is IMethodHandler chartOnHover && chartOnHover != null)
```
Comment (Joelius300) [inline src/ChartJs.Blazor/ChartJS/ChartJsInterop.cs]: Null checks can be left out because of the [is-operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is).

> The is expression is true if expr isn't null, and any of the following is true...

**Comment by Joelius300** [comment]
Investigate the `LegendItem.Index` property a bit more. It's not mentioned in [the docs](https://www.chartjs.org/docs/latest/configuration/legend.html#legend-item-interface) (at least not in the model; only the sample) but it obviously exists (right?) and also what is `DatasetIndex`? If it's the same as with [the tooltips](https://www.chartjs.org/docs/latest/configuration/tooltip.html#tooltip-item-interface), there's both?!

Make `BaseConfigOptions.Events` type safe with a string enum. Add a custom option like with `ChartType` for maybe custom events?

File: src/ChartJs.Blazor/Util/ExpandoObjectExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+using System.Dynamic;
+using System.Linq;
+using System.Text;
+
+namespace ChartJs.Blazor.Util
+{
+    /// <summary>
+    /// This class contains extensions for the <see cref="ExpandoObject"/> type which allows for easier access and modification particularly when nested.
+    /// </summary>
+    public static class ExpandoObjectExtensions
```
Comment (Joelius300) [inline src/ChartJs.Blazor/Util/ExpandoObjectExtensions.cs]: Should probably be internal

**Comment by Joelius300** [comment]
I have now done quite a bit of performance testing. The implementation as proposed by this PR is quite a bit slower than what we had before on certain callbacks so I investigated why. Turns out it's not the manual deserialization, it's not the dynamic invoking of a delegate, it's not the interop itself, no, it's the removal of circular references on JavaScript side which we need so `System.Text.Json` can parse it. Currently this implementation uses [this small extension](https://github.com/douglascrockford/JSON-js/blob/master/cycle.js) to remove the circular references. It's a recursive solution and when you have big objects like the _whole chart object_ that the legend filter receives ([source](https://www.chartjs.org/docs/latest/configuration/legend.html?h=and%20the%20chart%20data)), it starts to become really slow (thats also what I mentioned in the PR description about 'legend callbacks').

This means that extension has to go. It works fine but it doesn't suit our needs. The next place to look is probably [flatted](https://github.com/WebReflection/flatted). I'll try my best but as someone not very experienced with JavaScript, I'm very open for help on this 😅

**Comment by Joelius300** [comment]
From my side this PR is done. It won't be merged into master (!) but it will be the first big change for the upcoming 2.0 release and will be integrated there. There's a good change the merging has to be done by hand in order to preserve a good commit history. I'll do that once the changes are approved and the version stuff including milestones is done.

@mariusmuntean Could you please look over it? All the samples have been adjusted accordingly and the performance has been improved by a lot. You can test it by cloning my repo.

File: src/ChartJs.Blazor/Interop/DelegateHandler.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        public bool ReturnsValue => s_delegateHasReturnValue;
+
+        /// <summary>
+        /// Gets the indices of the ignored callback parameters. The parameters at these indices won't
+        /// be sent to C# and won't be deserialized. These indices are defined by the
+        /// <see cref="IgnoreCallbackValueAttribute"/>s on the delegate passed to this instance.
+        /// </summary>
+        public IReadOnlyCollection<int> IgnoredIndices { get; }
+
+        static DelegateHandler()
```
Comment (mariusmuntean) [inline src/ChartJs.Blazor/Interop/DelegateHandler.cs:52]: Am I right in assuming that the static constructor (and members) are a performance optimization, i.e. to avoid doing the same work multiple times for multiple instances of DelegateHandler when T is identical?

File: src/ChartJs.Blazor/Interop/ChartJsInterop.cs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+            // Get back an ExpandoObject dynamic with the clean config - having an ExpandoObject allows us to add/replace members regardless of type
+            ExpandoObject cleanChartConfig = JsonConvert.DeserializeObject<ExpandoObject>(cleanChartConfigStr, new ExpandoObjectConverter());
+
+#pragma warning disable CS0618 // Type or member is obsolete
+
+            // Restore any .net refs that need to be passed intact
+            // TODO Find a way to do this dynamically. Maybe with attributes or something like that?
+            dynamic dynamicChartConfig = (dynamic)chartConfig;
+            if (dynamicChartConfig?.Options?.OnClick is IMethodHandler chartOnClick)
```
[the remark this answers is not in the document]
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/ChartJsInterop.cs:144]: You mean putting them inside the `SetValue` directly? I thought it would make it a bit more clear like this.

File: src/ChartJs.Blazor/Interop/IMethodHandler.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    /// Represents a type that's able to handle method calls coming from JavaScript.
+    /// In order to maintain the strongly typed nature of C#, please prefer using <see cref="IMethodHandler{T}"/>.
+    /// </summary>
+    [Obsolete("Use " + nameof(IMethodHandler) + "<T> instead.")]
+    public interface IMethodHandler
+    {
+        /// <summary>
+        /// The name of the method which should be called from JavaScript.
+        /// The method name can be a reference to C# or JavaScript.
```
[the remark this answers is not in the document]
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/IMethodHandler.cs:30]: Agree, I'll make this more clear. As a side note, we probably won't keep that member in this interface when we separate the projects as this interface will become blazor independant and the method name is a blazor specific requirement.

File: src/ChartJs.Blazor/Interop/DelegateHandler.cs (hunk 5 above)
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/DelegateHandler.cs:52]: Correct. Performance isn't an important factor for our library yet but avoiding unnecessary computations (especially in reflection) is always a plus.

File: src/ChartJs.Blazor/Interop/ChartJsInterop.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mariusmuntean) [reply src/ChartJs.Blazor/Interop/ChartJsInterop.cs:144]: The `path` parameter of that extension method is already pretty well documented. Would you be willing to go for a named parameter syntax, a la `cleanChartConfig.SetValue(path: "options.onClick", chartOnClick);` ?

File: src/ChartJs.Blazor/Interop/DelegateHandler.cs (hunk 5 above)
Comment (mariusmuntean) [reply src/ChartJs.Blazor/Interop/DelegateHandler.cs:52]: Agreed!

File: src/ChartJs.Blazor/Interop/ChartJsInterop.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/ChartJsInterop.cs:144]: Sure, I'll do that.

File: src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts [hunk 8]
```diff
@@ -106,182 +100,194 @@ class ChartJsInterop {
...
-            if (typeof iClickHandler === "object" &&
-                iClickHandler.hasOwnProperty('fullFunctionName')) {
-                let onClickStringName: { fullFunctionName: string } = <any>iClickHandler;
-                const onClickNamespaceAndFunc = onClickStringName.fullFunctionName.split(".");
-                const onClickFunc = window[onClickNamespaceAndFunc[0]][onClickNamespaceAndFunc[1]];
-                if (typeof onClickFunc === "function") {
-                    return onClickFunc;
-                } else { // fallback to the default
-                    return chartJsDefaultHandler;
+    private GetMethodHandler(iMethodHandler, defaultFunc: Function) {
```
Comment (mariusmuntean) [inline src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts:216]: the type of iMethodHandler parameter could be `any`, avoiding the later cast.

File: src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts [hunk 9]
```diff
@@ -106,182 +100,194 @@ class ChartJsInterop {
...
+        if (!iMethodHandler ||
+            typeof iMethodHandler !== "object" ||
+            !iMethodHandler.hasOwnProperty('methodName')) {
+            return defaultFunc;
+        }
+
+        if (iMethodHandler.hasOwnProperty('handlerReference')) {
+            const handler: { handlerReference: DotNetObjectReference, methodName: string, returnsValue: boolean, ignoredIndices: number[] } = <any>iMethodHandler;
```
Comment (mariusmuntean) [inline src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts:224]: If you pull out the inlined type to an `interface` specification you can add a bit of documentation for its members and it becomes reusable, if needed later.

Comment (Joelius300) [reply src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts:224]: Good idea. That would also allow better ways of handling the casting and parameter.

File: src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts (hunk 8 above)
Comment (Joelius300) [reply src/ChartJs.Blazor/wwwroot/ts/ChartJsInterop.ts:216]: I think any not the call here but see the comment below; we can use other interfaces for both implementations of IMethodHandler`.

File: src/ChartJs.Blazor/Interop/ChartJsInterop.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/ChartJsInterop.cs:144]: I had missed it in the PR, it's been addressed in 4d99fa55f4.

File: src/ChartJs.Blazor/Interop/IMethodHandler.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Joelius300) [reply src/ChartJs.Blazor/Interop/IMethodHandler.cs:30]: I had missed it in the PR, it's been addressed in 4d99fa55f4.
